File: 351002/Zakrevskaya/src/user/user.service.ts
import { Injectable, NotFoundException, BadRequestException, HttpException, HttpStatus } from '@nestjs/common';
import { UserRepository } from './user.repository';
import { CreateUserDto, UserResponseDto } from './dto/user.dto';
import { User, UserRole } from './user.entity';
import * as bcrypt from 'bcrypt';

@Injectable()
export class UserService {
  constructor(private readonly userRepository: UserRepository) {}

  async create(createUserDto: CreateUserDto): Promise<UserResponseDto> {
    this.validateUserDto(createUserDto);

    const hashedPassword = await bcrypt.hash(createUserDto.password, 10);

    try {
      const user = await this.userRepository.create({
        login: createUserDto.login,
        password: hashedPassword,
        firstname: createUserDto.firstName,
        lastname: createUserDto.lastName,
        role: createUserDto.role || UserRole.CUSTOMER,
      });
      return this.toResponseDto(user);
    } catch (error: any) {
      if (error.code === '23505' || error.message?.includes('duplicate key')) {
        throw new HttpException('User with this login already exists', HttpStatus.FORBIDDEN);
      }
      throw error;
    }
  }

  async update(id: number, updateUserDto: CreateUserDto): Promise<UserResponseDto> {
    this.validateUserDto(updateUserDto);

    const hashedPassword = await bcrypt.hash(updateUserDto.password, 10);

    try {
      const user = await this.userRepository.update(id, {
        login: updateUserDto.login,
        password: hashedPassword,
        firstname: updateUserDto.firstName,
        lastname: updateUserDto.lastName,
        role: updateUserDto.role || UserRole.CUSTOMER,
      });
      return this.toResponseDto(user);
    } catch (error: any) {
      if (error.code === '23505' || error.message?.includes('duplicate key')) {
        throw new HttpException('User with this login already exists', HttpStatus.FORBIDDEN);
      }
      throw error;
    }
  }

  async delete(id: number): Promise<void> {
    const user = await this.userRepository.findById(id);
    if (!user) throw new NotFoundException('User not found');
    await this.userRepository.delete(id);
  }

  async findById(id: number): Promise<UserResponseDto> {
    const user = await this.userRepository.findById(id);
    if (!user) throw new NotFoundException('User not found');
    return this.toResponseDto(user);
  }

  async findAll(page: number = 1, limit: number = 10): Promise<UserResponseDto[]> {
    const [users] = await this.userRepository.findAll({ page, limit });
    return users.map(user => this.toResponseDto(user));
  }

  private validateUserDto(dto: CreateUserDto) {
    if (dto.login.length < 2 || dto.login.length > 64) {
      throw new BadRequestException('Login must be 2-64 characters');
    }
    if (dto.password.length < 8 || dto.password.length > 128) {
      throw new BadRequestException('Password must be 8-128 characters');
    }
    if (dto.firstName.length < 2 || dto.firstName.length > 64) {
      throw new BadRequestException('firstName must be 2-64 characters');
    }
    if (dto.lastName.length < 2 || dto.lastName.length > 64) {
      throw new BadRequestException('lastName must be 2-64 characters');
    }
  }

  private toResponseDto(user: User): UserResponseDto {
    return {
      id: user.id,
      login: user.login,
      firstName: user.firstname,
      lastName: user.lastname,
      role: user.role,
    };
  }
}

File: 351003/Razhnou/task350/publisher/src/main/java/by/bsuir/task350/publisher/service/UserService.java
package by.bsuir.task350.publisher.service;

import by.bsuir.task350.publisher.config.CacheNames;
import by.bsuir.task350.publisher.dto.request.UserRequestTo;
import by.bsuir.task350.publisher.dto.response.UserResponseTo;
import by.bsuir.task350.publisher.entity.User;
import by.bsuir.task350.publisher.exception.BadRequestException;
import by.bsuir.task350.publisher.exception.ConflictException;
import by.bsuir.task350.publisher.exception.NotFoundException;
import by.bsuir.task350.publisher.mapper.UserMapper;
import by.bsuir.task350.publisher.repository.UserRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
public class UserService {
    private final UserRepository userRepository;
    private final PublisherCacheService cacheService;

    public UserService(UserRepository userRepository, PublisherCacheService cacheService) {
        this.userRepository = userRepository;
        this.cacheService = cacheService;
    }

    @Transactional
    public UserResponseTo create(UserRequestTo request) {
        if (request.id() != null) {
            throw new BadRequestException("User id must be null on create", 3);
        }
        validateLogin(request.login());
        validatePassword(request.password());
        validateHumanName(request.firstname(), "User firstname");
        validateHumanName(request.lastname(), "User lastname");
        validateUniqueLoginOnCreate(request.login());

        User user = UserMapper.toEntity(request);
        UserResponseTo response = UserMapper.toResponse(userRepository.save(user));
        cacheService.put(CacheNames.USERS, response.id(), response);
        return response;
    }

    @Transactional(readOnly = true)
    public List<UserResponseTo> findAll() {
        return userRepository.findAll().stream()
                .map(UserMapper::toResponse)
                .toList();
    }

    @Transactional(readOnly = true)
    public UserResponseTo findById(Long id) {
        validateId(id, "User id");
        UserResponseTo cached = cacheService.get(CacheNames.USERS, id, UserResponseTo.class);
        if (cached != null) {
            return cached;
        }
        UserResponseTo response = UserMapper.toResponse(getUser(id));
        cacheService.put(CacheNames.USERS, id, response);
        return response;
    }

    @Transactional
    public UserResponseTo update(UserRequestTo request) {
        validateId(request.id(), "User id");
        validateLogin(request.login());
        validatePassword(request.password());
        validateHumanName(request.firstname(), "User firstname");
        validateHumanName(request.lastname(), "User lastname");
        validateUniqueLoginOnUpdate(request.login(), request.id());

        User user = getUser(request.id());
        UserMapper.updateEntity(user, request);
        UserResponseTo response = UserMapper.toResponse(userRepository.save(user));
        cacheService.put(CacheNames.USERS, response.id(), response);
        return response;
    }

    @Transactional
    public void delete(Long id) {
        validateId(id, "User id");
        userRepository.delete(getUser(id));
        cacheService.evict(CacheNames.USERS, id);
    }

    public User getUser(Long id) {
        return userRepository.findById(id)
                .orElseThrow(() -> new NotFoundException("User not found", 1));
    }

    private void validateUniqueLoginOnCreate(String login) {
        if (userRepository.existsByLogin(login.trim())) {
            throw new ConflictException("User login already exists", 1);
        }
    }

    private void validateUniqueLoginOnUpdate(String login, Long id) {
        if (userRepository.existsByLoginAndIdNot(login.trim(), id)) {
            throw new ConflictException("User login already exists", 1);
        }
    }

    private void validateId(Long id, String fieldName) {
        if (id == null || id <= 0) {
            throw new BadRequestException(fieldName + " must be greater than 0", 1);
        }
    }

    private void validateLogin(String login) {
        validateTextLength(login, "User login", 2, 64, 2);
    }

    private void validatePassword(String password) {
        validateTextLength(password, "User password", 8, 128, 4);
    }

    private void validateHumanName(String value, String fieldName) {
        validateTextLength(value, fieldName, 2, 64, 5);
    }

    private void validateTextLength(String value, String fieldName, int min, int max, int suffix) {
        if (value == null || value.trim().isEmpty()) {
            throw new BadRequestException(fieldName + " must not be blank", suffix);
        }
        int length = value.trim().length();
        if (length < min || length > max) {
            throw new BadRequestException(fieldName + " length must be between " + min + " and " + max, suffix);
        }
    }
}
